pkg/responsemeta/responsemeta.go

```diff
@@ -0,0 +1,111 @@
+package responsemeta
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/rs/zerolog/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	dispatch "github.com/authzed/spicedb/internal/proto/dispatch/v1"
+)
+
+// NewHandler creates a new response metadata handler for setting the response
+// metadata into the trailer of an API request.
+func NewHandler(ctx context.Context) *ResponseMetadataHandler {
+	return &ResponseMetadataHandler{
+		ctx:    ctx,
+		wasSet: false,
+	}
+}
+
+// ResponseMetadataHandler is a helper struct for tracking whether response
+// metadata was written to the trailer.
+type ResponseMetadataHandler struct {
+	ctx    context.Context
+	wasSet bool
+}
+
+// SetFor sets the response metadata in the trailer to that contained in the given
+// dispatch metadata.
+func (rmh *ResponseMetadataHandler) SetFor(dispatchMetadata *dispatch.ResponseMeta) {
+	err := setResponseMetadata(rmh.ctx, dispatchMetadata)
+	if err != nil {
+		log.Error().Err(err).Msg("error setting response metadata")
+		return
+	}
+	rmh.wasSet = true
+}
+
+// SetDefaultsIfNotCalled sets default response metadata if and only if SetFor was
+// not already called.
+func (rmh *ResponseMetadataHandler) SetDefaultsIfNotCalled() {
+	if !rmh.wasSet {
+		err := SetDefaultResponseMetadata(rmh.ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("error setting default response metadata")
+			return
+		}
+		rmh.wasSet = true
+	}
+}
+
+// ResponseMetadataKey defines a key in the response metadata trailer.
+type ResponseMetadataKey string
+
+const (
+	// DispatchedOperationsCount is the key in the response trailer metadata for
+	// the number of dispatched operations that were needed to perform the overall
+	// API call.
+	DispatchedOperationsCount ResponseMetadataKey = "io.spicedb.respmeta.dispatchedoperationscount"
+
+	// CachedOperationsCount is the key in the response trailer metadata for
+	// the number of *cached* operations that would have been otherwise dispatched
+	// to perform the overall API call.
+	CachedOperationsCount ResponseMetadataKey = "io.spicedb.respmeta.cachedoperationscount"
+)
+
+// setResponseMetadata sets the external response metadata, extracted from the dispatch
+// response metadata, on the given context.
+func setResponseMetadata(ctx context.Context, dispatchMetadata *dispatch.ResponseMeta) error {
+	trailer := metadata.Pairs(
+		// DispatchedOperationsCount
+		string(DispatchedOperationsCount),
+		fmt.Sprintf("%d", dispatchMetadata.DispatchCount),
```

Comment:
`strconv.Itoa`?

Reply:
I tend to prefer using `fmt` for converting to strings, but I can switch if you like

Reply:
https://gist.github.com/evalphobia/caee1602969a640a4530

Reply:
Updated